File: iOS/PricetarPartner/PricetarPartner/Http/Requester/FetchUserRequester.swift
//
//  FetchUserRequester.swift
//  PricetarPartner
//
//  Created by Leapfrog-Inc. on 2021/02/21.
//

import Foundation

class FetchUserRequester {
    
    static let shared = FetchUserRequester()
    
    var dataList = [UserData]()
    
    func fetch(completion: @escaping ((Bool) -> ())) {
        
        var params = ["command": "getUser"]
        params["userId"] = SaveData.shared.userId
        
        ApiRequester.post(params: params) { [weak self] result, data in
            if result, let dic = data as? Dictionary<String, Any> {
                if (dic["result"] as? String) == "0", let users = (dic["users"] as? Array<Dictionary<String, Any>>) {
                    self?.dataList = users.compactMap { UserData(data: $0) }
                    completion(true)
                    return
                }
            }
            completion(false)
        }
    }
    
    func query(userId: String) -> UserData? {
        return self.dataList.first { $0.id == userId }
    }
}

File: Android/app/src/main/java/jp/co/lfg/pricetarpartner/Http/Requester/FetchUserRequester.java
package jp.co.lfg.pricetarpartner.Http.Requester;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

import jp.co.lfg.pricetarpartner.Http.ApiRequester;
import jp.co.lfg.pricetarpartner.Http.DataModel.UserData;
import jp.co.lfg.pricetarpartner.System.SaveData;

public class FetchUserRequester {

    private static FetchUserRequester mRequester = new FetchUserRequester();

    private FetchUserRequester() {}

    public static FetchUserRequester getInstance() {
        return mRequester;
    }

    public ArrayList<UserData> dataList = new ArrayList<>();

    public void fetch(final Callback callback) {

        HashMap<String, String> params = new HashMap<>();
        params.put("command", "getUser");
        params.put("userId", SaveData.getInstance().userId);

        ApiRequester.post(params, new ApiRequester.Callback() {
            @Override
            public void didReceiveData(boolean result, JSONObject json) {
                if (result) {
                    try {
                        String apiResult = json.getString("result");
                        if (apiResult.equals("0")) {
                            ArrayList<UserData> users = new ArrayList<>();
                            JSONArray userArray = json.getJSONArray("users");
                            for (int i = 0; i < userArray.length(); i++) {
                                UserData userData = UserData.create(userArray.getJSONObject(i));
                                if (userData != null) {
                                    users.add(userData);
                                }
                            }
                            dataList = users;
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch (JSONException e) {}
                }
                callback.didReceiveData(false);
            }
        });
    }

    public UserData query(String userId) {

        for (UserData userData : dataList) {
            if (userData.id.equals(userId)) {
                return userData;
            }
        }
        return null;
    }

    public interface Callback {
        void didReceiveData(boolean result);
    }
}
